Feat: Move `FlatList`s to `@react-native/flat-lists` package

Pranav-yadav opened the pull request:
## Summary

This PR moves `FlatList`, `SectionList` and related files _(tests, etc.)_ to a new package `@react-native/flat-lists` located under `packages/flat-lists` as proposed in #35263
This is a first step towards moving all the `FlatList`-related code to a separate package.
This will allow us to make changes to the `FlatList` implementation without affecting the rest of the React Native codebase.

## Changelog

[General] [Changed] - Move `FlatList`s to `@react-native/flat-lists` package.
`FlatList` and `SectionList` are now available in `@react-native/flat-lists` package.
**Warning**: This is a breaking(?) change, and you may need to update your imports to use the new package.

## Test Plan

- ensure `yarn lint && yarn flow && yarn test` is `#00ff00`

Pranav-yadav commented:
Depends on #35406 

PS: do _tag me_ in comments/reply

TODOs: Once the dependent PR is merged;

- [ ] Rebase 
- [ ] rm files related to `VirtualizedLists` and the helpers
- [ ] Update imports for/to  `@react-native/virtualized-lists`
- [ ] Fix CI (flow)

<details><summary>Tests</summary>

![image](https://user-images.githubusercontent.com/55224033/203132288-0aaa3c32-da47-479e-9e02-2663b94f928a.png)

Except Flow :)
</details>

necolas reviewed (changes requested):
Have a look at the feedback on the VirtualizedList PR. Specifically:

* We should not be reaching inside any packages. Instead `@react-native/flat-lists` should export everything that can be imported, and it should import from `react-native` and not internal paths.
* This package should have the virtualized-lists package as a dependency, rather than creating circular dependencies.

https://github.com/facebook/react-native/pull/35406

cortinico commented on packages/flat-lists/package.json:
```diff
@@ -0,0 +1,17 @@
+{
+    "name": "@react-native/flat-lists",
+    "version": "1.0.0",
```
Same as the other package. Versioning here should be updated to `0.72.0`

Pranav-yadav replied:
done

cortinico commented on packages/flat-lists/package.json:
```diff
@@ -0,0 +1,17 @@
+    },
+    "license": "MIT",
+    "peerDependencies": {
+      "react": "18.2.0",
+      "react-native": "*",
+      "@react-native/virtualized-lists": "1.0.0",
+      "react-test-renderer": "18.2.0"
```
```suggestion
      "react-native": "*",
      "@react-native/virtualized-lists": "1.0.0",
```

Or do you have a specific reason to have a peerDependency on react?

Pranav-yadav replied:
frankly, not sure.. interacting 1st time with these libraries :)
`react` is used in `FlatList`s components as well as related tests, if it was separate package (repo) itself I would've kept it, not sure how deps scoping and resolving will work in monorepo (?)

cortinico replied:
I believe you should be fine by setting the peerDependency on RN only. RN is providing react for you and you'll be good. Specifying a React version here adds unnecessary complexity for us.

Pranav-yadav replied:
Thanks for letting me know :)

Pranav-yadav commented:
@gabrieldonadel can you help me with these https://app.circleci.com/pipelines/github/facebook/react-native/17862/workflows/19e750f6-a614-4430-b201-aadc52fd0558/jobs/366958?invite=true#step-108-71,? (_expect those which are related to `@react-native/virtualized-lists`_)

Tests are failing, even though they're exported here `packages/flat-lists/Lists/SectionList.d.ts`

Pranav-yadav commented:
- rebased (reverted Verdaccio config..)
- fixed errors except those related to `@react-native/virtualized-lists`

gabrieldonadel commented:
@Pranav-yadav now that https://github.com/facebook/react-native/commit/2e3dbe9c2fbff52448e2d5a7c1e4c96b1016cf25 was merged can you rebase this from main?

Pranav-yadav commented:
> @Pranav-yadav now that 2e3dbe9 was merged can you rebase this from main?

Sure. Thanks for the ping.

Pranav-yadav commented:
@NickGerleman / @hoxyq I have rebased this PR on main as the `@react-native/virtualized-lists` commit were merged.

Update:
- rebased on main
- fixed some typescripts errors

cc: @necolas
PS: Since this PR is being tracked under umbrella issue(s) you may add `Tech: Monorepo` label ...

Pranav-yadav commented:
@necolas re-pinging, in case you missed.. 🙂

Pranav-yadav commented:
@NickGerleman please let me know if should rebase again so that you can review it afterwards.

necolas commented on Libraries/Animated/components/AnimatedFlatList.js:
```diff
@@ -10,8 +10,8 @@
 
-import FlatList from '../../Lists/FlatList';
 import createAnimatedComponent from '../createAnimatedComponent';
+import {FlatList} from '@react-native/flat-lists';
```
What we did with virtualized list is kept all the existing files and imported the exports from the new package. That way none of the internal paths in RN change, and imports outside of FlatList won't be to be rewritten.

Pranav-yadav replied:
Sure. Will rebase and try to keep it that way...

ps: Thanks for reply :)

necolas replied:
If it helps you can refer to the virtualized-lists package PR: https://github.com/facebook/react-native/pull/35406

Pranav-yadav replied:
Rebased on the main and updated.

Pranav-yadav commented on packages/flat-lists/Lists/FlatList.js.flow:
```diff
@@ -24,17 +24,17 @@ import {typeof VirtualizedList} from '@react-native/virtualized-lists';
-   * immutable list, use the underlying `VirtualizedList` directly.
+   * An array (or array-like list) of items to render. Other data types can be
+   * used by targetting VirtualizedList directly.
    */
-  data: ?$ReadOnlyArray<ItemT>,
+  data: ?$ArrayLike<ItemT>,
 |};
 type OptionalProps<ItemT> = {|
   renderItem?: ?RenderItemType<ItemT>,
   columnWrapperStyle?: ViewStyleProp,
   extraData?: any,
   getItemLayout?: (
-    data: ?Array<ItemT>,
+    data: ?$ArrayLike<ItemT>,
```
This was left from the recent PR by Nick.

Pranav-yadav replied:
he fixed this afterwards..

Pranav-yadav commented:
<details><summary>Tests Results</summary>

`yarn run flow`
![image](https://user-images.githubusercontent.com/55224033/226872011-ea6c932c-797d-4658-8271-91fc6519a81a.png)

`yarn jest flat-lists`
![image](https://user-images.githubusercontent.com/55224033/226877275-f71475e2-0b43-42f9-ad74-a8b0c8db99f6.png)

`yarn test-ci`
![image](https://user-images.githubusercontent.com/55224033/226877455-feb6470a-7808-41ec-a87b-196b98b67b5d.png)

`yarn lint`
Doesn't works on my Windows Machine :(
</details>

Pranav-yadav commented on packages/flat-lists/package.json:
```diff
@@ -0,0 +1,27 @@
+      "test": "jest"
+    },
+    "license": "MIT",
+    "dependencies": {
+      "@react-native/virtualized-lists": "^0.72.2",
+      "invariant": "^2.2.4",
+      "memoize-one": "^5.0.0"
+    },
+    "devDependencies": {
+      "react-test-renderer": "18.2.0"
+    },
+    "peerDependencies": {
+      "react-native": "*",
+      "react-test-renderer": "18.2.0"
+    }
```
I think there's definitely something wrong here. So, instead of me wasting precious CI compute, may be anyone can help me figure this out??

Though, Nicola had commented on the same thing earlier, but that was failing ci, as there was no monorepo setup at that time..

Pranav-yadav replied:
false positive :)

NickGerleman commented:
Sorry I haven't been able to take a close look at this (though it seems like you found some help on the Discord). Though as a general bit of feedback, I'm not sure it makes sense to add a new plural `flat-lists` package, when we have one FlatList, vs adding to the existing plural `virtualized-lists` package, which currently also only has one list.

necolas commented:
They both also have the section list variant

Pranav-yadav commented:
>Sorry I haven't been able to take a close look at this.

No worries :)

>I'm not sure it makes sense to add a new **_plural_** flat-lists package.

I did consider that at the start only but, it's not a `FlatList` only package; `virtualized-lists` and `flat-lists` have `SectionList` variants.
and specifically `flat-lists` has:
1. `SectionList`, 
2. `SectionListModern`

variants as well.. as Necolas said.
So I think `plural` seems logical here, yet it should have been singular only, iff `SectionList` had been isolated earlier...

>They both also have the section list variant

Yeah.

Pranav-yadav commented:
# 🎉 All Green !!

### Will anyone import this?
@NickGerleman

necolas commented on packages/flat-lists/Lists/FlatList.js:
```diff
@@ -0,0 +1,700 @@
+ */
+
+import typeof ScrollViewNativeComponent from 'react-native';
+import type {ViewStyleProp} from 'react-native/Libraries/StyleSheet/StyleSheet';
```
Pleas can we avoid having any imports from RN internals

Pranav-yadav replied:
Yeah. I forgot that; this way it may not work in outside projects.

Right now I'm in college, will try to change it once I reach home.

Thanks.

Pranav-yadav replied:
Sorry for the delay. I'm busy with assignment tests at college, will try to look at this; tomorrow eod..

roryabraham commented:
Apologies if this has been discussed already, but might it be simpler to just include these files in the @react-native/virtualized-lists package? Is there any advantage to keeping them separate?

Pranav-yadav commented:
Theoretically "Yes". But, both lists were implemented and exist for a _special_ (different) _purpose_, keeping them separate would support the _initial cause of "separation of concerns"_.

If we combine them into a "single" package:
1. Sine, both have SectionList variants it could be confusing.
2. Would significantly increase the package size. (unwanted)
3. There is _possibility_ that some types may collide.

WDYT?

P.S.: Personally, I would've loved to have them in a single package named `@react-native/lists` (or similar) _in the first place_ but, doing this change now would need a breaking release while keeping it backward compatible.
